python/strings_udf/cpp/include/cudf/strings/udf/split.cuh

```diff
@@ -0,0 +1,192 @@
+    auto const length         = (pos < 0 ? nchars : pos) - last_pos;
+    if (result) { *result++ = d_str.substr(last_pos, length); }
+    last_pos = pos + tgt.length();
+    ++count;
+    if (pos < 0) { break; }
+  }
+
+  return count;
+}
+
+/**
+ * @brief Split string using given target array
+ *
+ * @param d_str String to split
+ * @param tgt Character array encoded in UTF-8 used for identifying split points
+ * @param bytes Number of bytes to read from `tgt`
+ * @param result Empty array to populate with output objects.
+ *               Pass `nullptr` to just get the token count.
+ * @return Number of tokens returned
+ */
+__device__ inline int split(cudf::string_view const d_str,
+                            char const* tgt,
+                            cudf::size_type bytes,
+                            cudf::string_view* result)
+{
+  return split(d_str, cudf::string_view{tgt, bytes}, result);
+}
+
+/**
+ * @brief Split string using given target array
+ *
+ * @param d_str String to split
+ * @param tgt Null-terminated character array encoded in UTF-8 used for identifying split points
+ * @param result Empty array to populate with output objects.
+ *               Pass `nullptr` to just get the token count.
+ * @return Number of tokens returned
+ */
+__device__ inline int split(cudf::string_view const d_str,
+                            char const* tgt,
+                            cudf::string_view* result)
+{
+  return split(d_str, tgt, detail::bytes_in_null_terminated_string(tgt), result);
+}
+
+/**
+ * @brief Split string on whitespace
+ *
+ * This will create tokens by splitting on one or more consecutive whitespace characters
+ * found in `d_str`.
+ *
+ * @param d_str String to split
+ * @param result Empty array to populate with output objects.
+ *               Pass `nullptr` to just get the token count.
+ * @return Number of tokens returned
+ */
+__device__ inline cudf::size_type split(cudf::string_view const d_str, cudf::string_view* result)
+{
+  cudf::strings::detail::whitespace_string_tokenizer tokenizer{d_str};
+  cudf::size_type count = 0;
+  while (tokenizer.next_token()) {
+    auto token = tokenizer.get_token();
+    if (result) { *result++ = d_str.substr(token.first, token.second - token.first); }
+    ++count;
+  }
+  return count;
+}
+
+/**
+ * @brief Join an array of strings with a separator
+ *
+ * @code{.cpp}
+ * auto separator = cudf::string_view{"::", 2};
+ * cudf::string_view input[] = {
+ *   cudf::string_view{"hello", 5},
+ *   cudf::string_view{"goodbye", 7},
+ *   cudf::string_view{"world", 5} };
+ *
+ * auto result = join(separator, input, 3);
+ * // result is "hello::goodbye::world"
+ * @endcode
+ *
+ * @param separator Separator string
+ * @param input An array of strings to join
+ * @param count Number of elements in `input`
+ * @return New string
+ */
+__device__ inline udf_string join(cudf::string_view const separator,
+                                  cudf::string_view* input,
```

Comment:
Should we be using device spans instead of passing both pointers and lengths? Might be a limitation in the calling code?

Reply:
Maybe. Perhaps we can revisit this when the cython code is added?